lib/scholar/cluster/gmm.ex

```diff
@@ -0,0 +1,407 @@
+
+    weights = weights / Nx.sum(weights)
+    {weights, means, covariances, precisions_cholesky}
+  end
+
+  defnp estimate_gaussian_parameters(x, resp, reg_covar) do
+    nk = Nx.sum(resp, axes: [0])
+    means = Nx.dot(Nx.transpose(resp), x) / nk
+    covariances = estimate_gaussian_covariances(x, resp, nk, means, reg_covar)
+    precisions_cholesky = compute_precisions_cholesky(covariances)
+    {nk, means, covariances, precisions_cholesky}
+  end
+
+  defnp estimate_gaussian_covariances(x, resp, nk, means, reg_covar) do
+    type = Nx.type(x)
+    {num_gaussians, num_features} = Nx.shape(means)
+
+    {covariances, _} =
+      while {
+              covariances =
+                Nx.tensor(0.0, type: type)
+                |> Nx.broadcast({num_gaussians, num_features, num_features}),
+              {k = 0, x, resp, nk, means, reg_covar}
+            },
+            k < num_gaussians do
+        diff = x - means[k]
+        covariance = Nx.dot(resp[[.., k]] * Nx.transpose(diff), diff) / nk[k]
+        covariance = Nx.put_diagonal(covariance, Nx.take_diagonal(covariance) + reg_covar)
+
+        {
+          Nx.put_slice(covariances, [k, 0, 0], Nx.new_axis(covariance, 0)),
+          {k + 1, x, resp, nk, means, reg_covar}
+        }
+      end
+
+    covariances
+  end
+
+  defnp compute_precisions_cholesky(covariances) do
+    type = Nx.type(covariances)
+    {num_gaussians, num_features, _} = Nx.shape(covariances)
+
+    {precisions_cholesky, _} =
+      while {
+              precisions_cholesky =
+                Nx.tensor(0.0, type: type)
+                |> Nx.broadcast({num_gaussians, num_features, num_features}),
+              {k = 0, covariances}
+            },
+            k < num_gaussians do
+        covariance_cholesky = Nx.LinAlg.cholesky(covariances[k])
+
+        precision_cholesky =
+          covariance_cholesky
+          |> Nx.LinAlg.triangular_solve(Nx.eye(num_features, type: type), lower: true)
+          |> Nx.transpose()
+
+        {
+          Nx.put_slice(precisions_cholesky, [k, 0, 0], Nx.new_axis(precision_cholesky, 0)),
+          {k + 1, covariances}
+        }
+      end
+
+    precisions_cholesky
+  end
+
+  defnp estimate_log_prob_resp(x, weights, means, precisions_cholesky) do
+    weighted_log_prob = estimate_weighted_log_prob(x, weights, means, precisions_cholesky)
+    log_prob_norm = logsumexp(weighted_log_prob)
+    log_resp = weighted_log_prob - Nx.new_axis(log_prob_norm, 1)
+    {log_prob_norm, log_resp}
+  end
+
+  defnp estimate_weighted_log_prob(x, weights, means, precisions_cholesky) do
+    estimate_log_prob(x, means, precisions_cholesky) + Nx.log(weights)
+  end
+
+  defnp estimate_log_prob(x, means, precisions_cholesky) do
+    type = Nx.type(x)
+    {num_samples, num_features} = Nx.shape(x)
+    {num_gaussians, _} = Nx.shape(means)
+    log_det = compute_log_det_cholesky(precisions_cholesky)
+
+    {log_prob, _} =
+      while {
```

Comment:
vectorization too

Reply:
Yeah, also doesn't seem hard to vectorize.